cabal-install/Distribution/Client/Dependency/Modular/Linking.hs

```diff
@@ -216,37 +224,46 @@ pickConcrete qpn@(Q pp _) i = do
       Just lg ->
         makeCanonical lg qpn i
 
-pickLink :: QPN -> I -> PP -> FlaggedDeps comp QPN -> UpdateState ()
-pickLink qpn@(Q pp pn) i pp' deps = do
+pickLink :: QPN -> I -> PP -> FlaggedDeps Component QPN -> UpdateState ()
+pickLink qpn@(Q _pp pn) i pp' deps = do
     vs <- get
-    -- Find the link group for the package we are linking to, and add this package
+
+    -- The package might already be in a link group
+    -- (because one of its reverse dependencies is)
+    let lgSource = case M.lookup qpn (vsLinks vs) of
+                     Nothing -> lgSingleton qpn Nothing
+                     Just lg -> lg
+
+    -- Find the link group for the package we are linking to
     --
     -- Since the builder never links to a package without having first picked a
     -- concrete instance for that package, and since we create singleton link
     -- groups for concrete instances, this link group must exist (and must
     -- in fact already have a canonical member).
-    let lg = vsLinks vs ! Q pp' pn
+    let target   = Q pp' pn
+        lgTarget = vsLinks vs ! Q pp' pn
 
     -- Verify here that the member we add is in fact for the same package and
     -- matches the version of the canonical instance. However, violations of
     -- these checks would indicate a bug in the linker, not a true conflict.
     let sanityCheck :: Maybe (PI PP) -> Bool
         sanityCheck Nothing              = False
-        sanityCheck (Just (PI _ canonI)) = pn == lgPackage lg && i == canonI
-    assert (sanityCheck (lgCanon lg)) $ return ()
+        sanityCheck (Just (PI _ canonI)) = pn == lgPackage lgTarget && i == canonI
+    assert (sanityCheck (lgCanon lgTarget)) $ return ()
+
+    -- Merge the two link groups (updateLinkGroup will propagate the change)
+    lgTarget' <- lift' $ lgMerge [] lgSource lgTarget
+    updateLinkGroup lgTarget'
 
-    -- Since we already have a canonical member, we just need to add the new
-    -- member into the group
-    let lg' = lg { lgMembers = S.insert pp (lgMembers lg) }
-    updateLinkGroup lg'
-    linkDeps [P qpn] pp' deps
+    -- Make sure all dependencies are linked as well
+    linkDeps target [P qpn] deps
```

Comment:
Ok, this still makes me feel uneasy. @grayjay, @edsko, am I right that the correctness of this relies on the SIR being enforced? I would normally expect that we have to include both `P qpn` and `P target` in the blame set.

Consider the following database:
```
db22 :: ExampleDb
db22 = [
    Right $ exAv "X" 1 [ExFix "C" 2, ExAny "A"]
  , Right $ exAv "Y" 1 [ExFix "C" 1, ExFix "A" 2]
  , Right $ exAv "A" 1 []
  , Right $ exAv "A" 2 [ExAny "B"]
  , Right $ exAv "B" 1 [ExAny "C"]
  , Right $ exAv "C" 1 []
  , Right $ exAv "C" 2 []
  ]
```
and let's assume we solve for `0.X` and `1.Y`. We get the following trace:
```
[__0] trying: 0.X-1.0.0 (user goal)
[__1] trying: 0.A-2.0.0 (dependency of 0.X-1.0.0)
[__2] trying: 0.B-1.0.0 (dependency of 0.A-2.0.0)
[__3] trying: 0.C-2.0.0 (dependency of 0.X-1.0.0)
[__4] trying: 1.Y-1.0.0 (user goal)
[__5] trying: 1.A~>0.A-2.0.0 (dependency of 1.Y-1.0.0)
[__6] trying: 1.B~>0.B-1.0.0 (dependency of 1.A-2.0.0)
[__7] next goal: 1.C (dependency of 1.Y-1.0.0)
[__7] rejecting: 1.C~>0.C-2.0.0 (conflict: 1.Y => 1.C==1.0.0)
[__7] rejecting: 1.C-2.0.0 (multiple instances)
[__7] rejecting: 1.C-1.0.0 (dependencies not linked: cannot make 1.C canonical member of {*0.C-2.0.0,1.C-2.0.0}; conflict set: 0.C, 1.B, 1.C)
[__7] fail (backjumping, conflict set: 0.C, 1.B, 1.C, 1.Y)
[__6] rejecting: 1.B-1.0.0 (multiple instances)
[__6] fail (backjumping, conflict set: 0.B, 0.C, 1.A, 1.B, 1.C, 1.Y)
[__5] rejecting: 1.A-2.0.0 (multiple instances)
[__5] rejecting: 1.A-1.0.0 (conflict: 1.Y => 1.A==2.0.0)
[__4] fail (backjumping, conflict set: 0.A, 0.B, 0.C, 1.A, 1.B, 1.C, 1.Y)
[__3] rejecting: 0.C-1.0.0 (conflict: 0.X => 0.C==2.0.0)
[__2] fail (backjumping, conflict set: 0.A, 0.B, 0.C, 0.X, 1.A, 1.B, 1.C, 1.Y)
[__1] trying: 0.A-1.0.0
[__2] trying: 0.C-2.0.0 (dependency of 0.X-1.0.0)
[__3] trying: 1.Y-1.0.0 (user goal)
[__4] next goal: 1.A (dependency of 1.Y-1.0.0)
[__4] rejecting: 1.A~>0.A-1.0.0 (conflict: 1.Y => 1.A==2.0.0)
[__4] trying: 1.A-2.0.0
[__5] trying: 1.B-1.0.0 (dependency of 1.A-2.0.0)
[__6] next goal: 1.C (dependency of 1.Y-1.0.0)
[__6] rejecting: 1.C~>0.C-2.0.0 (conflict: 1.Y => 1.C==1.0.0)
[__6] rejecting: 1.C-2.0.0 (multiple instances)
[__6] trying: 1.C-1.0.0
[__7] done
```

The interesting place is on level 7, where we reject all options for `C` and then backjump with the conflict set `{0.C, 1.B, 1.C, 1.Y}` which does *not* contain `0.B`.

But `0.B` in the conflict set is ultimately necessary in order to make us reconsider `0.A`, which was the choice that really needs to be fixed.

We do pick up `0.B` here, but only on level 6, because of
```
[__6] rejecting: 1.B-1.0.0 (multiple instances)
[__6] fail (backjumping, conflict set: 0.B, 0.C, 1.A, 1.B, 1.C, 1.Y)
```
This is because for the computation of the SIR conflict set, we indeed include both the link source and the link target.

I am, however, somewhat reluctant of relying on the presence of the SIR for the correctness of the solver. After all, it's possible we will remove or relax it at some point in the future.


The reason is that it's quite possible that we discover a failure underneath the current node that cannot be fixed by changing